Improve kill effect

EquipableMC opened the pull request:
### Description
This simply improves the kill effect by adding an option to prevent totems from popping

---
**Target Minecraft Versions:** any
**Requirements:** none
**Related Issues:** none

Efnilite commented on src/main/java/ch/njol/skript/effects/EffKill.java:
```diff
@@ -73,11 +75,14 @@ protected void execute(final Event e) {
 			}
 
+			if (IGNORE_TOTEM) {
+				HealthUtils.setHealth((Damageable) entity, 0);
```
missing return? or else you set health to 0 and then do damage, or intended?

sovdeeth replied:
why was this resolved?

EquipableMC replied:
I lowkey don't know why, I must of forgotten to push changes the first time... oops

APickledWalrus commented on src/main/java/ch/njol/skript/effects/EffKill.java:
```diff
@@ -44,24 +44,26 @@
 
 	static {
-		Skript.registerEffect(EffKill.class, "kill %entities%");
+		Skript.registerEffect(EffKill.class, "kill %entities%", "kill %entities% ignoring [the] (resurrection|revival|[undying] totem|totem [of undying])");
```
I'm not a huge fan of the syntax. I might prefer to just keep it simple:
```
ignoring [the|any] totem[s] [of undying]
```
I'd like to see what others think too though

EquipableMC replied:
What me and others have discussed on discord, is that the kill effect should be recoded to actually kill, since as of right now, it really just feels like a hardcoded damage effect. Another point is that the vanilla /kill command ignores the totem, whereas this, it does _not_ ignore the totem.

sovdeeth replied:
Pickle is talking about the pattern being overly complex, not about it's behavior.

Efnilite replied:
personally i think `ignoring [the|any] totem[s] of undying` should be it, as `ignoring totems` sounds too ambiguous imo

sovdeeth commented on src/main/java/ch/njol/skript/effects/EffKill.java:
```diff
@@ -44,24 +44,26 @@
-		"kill all endermen, witches and bats"})
+		"kill all endermen, witches and bats",
+		"kill the player ignoring totem of undying",
+		"kill target entity ignoring the totem",
```
agreed

Efnilite commented on src/main/java/ch/njol/skript/effects/EffKill.java:
```diff
@@ -75,7 +80,7 @@ protected void execute(final Event e) {
 	
 	@Override
-	public String toString(final @Nullable Event e, final boolean debug) {
+	public String toString(@Nullable Event e, boolean debug) {
 		return "kill" + entities.toString(e, debug);
```
```suggestion
	public String toString(@Nullable Event event, boolean debug) {
		return "kill " + entities.toString(event, debug);
```

Efnilite replied:
missing space

Efnilite commented on src/main/java/ch/njol/skript/effects/EffKill.java:
```diff
@@ -26,40 +26,45 @@
+		"kill all endermen, witches and bats",
+		"kill the player ignoring totem of undying",
+		"kill target entity ignoring the totem",
+		"kill target player ignoring resurrection"})
```
```suggestion
		"kill target player ignoring resurrection"
})
```

Efnilite commented on src/main/java/ch/njol/skript/effects/EffKill.java:
```diff
@@ -26,40 +26,45 @@
-@Since("1.0")
+		"kill all endermen, witches and bats",
+		"kill the player ignoring totem of undying",
+		"kill target entity ignoring the totem",
```
```suggestion
		"kill target entity ignoring totem of undying",
```

Efnilite commented on src/main/java/ch/njol/skript/effects/EffKill.java:
```diff
@@ -26,40 +26,45 @@
 
 	static {
-		Skript.registerEffect(EffKill.class, "kill %entities%");
+		Skript.registerEffect(EffKill.class, "kill %entities%", "kill %entities% ignoring [the|any] totem[s] of undying");
```
```suggestion
		Skript.registerEffect(EffKill.class, "kill %entities%", "kill %entities% ignoring [a|any] totem[s] of undying");
```

Efnilite commented on src/main/java/ch/njol/skript/effects/EffKill.java:
```diff
@@ -26,40 +26,45 @@
 
 	@Override
-	protected void execute(final Event e) {
+	protected void execute(Event e) {
```
```suggestion
	protected void execute(Event event) {
```

Efnilite commented on src/main/java/ch/njol/skript/effects/EffKill.java:
```diff
@@ -26,40 +26,46 @@
+	protected void execute(Event event) {
+		for (Entity entity : entities.getArray(event)) {
 
 			if (entity instanceof EnderDragonPart) {
 				entity = ((EnderDragonPart) entity).getParent();
```
```suggestion
			if (entity instanceof EnderDragonPart part) {
				entity = part.getParent();
```

Efnilite commented on src/main/java/ch/njol/skript/effects/EffKill.java:
```diff
@@ -26,40 +26,46 @@
-				final boolean creative = entity instanceof Player && ((Player) entity).getGameMode() == GameMode.CREATIVE;
+				boolean creative = entity instanceof Player player && player.getGameMode() == GameMode.CREATIVE;
 				if (creative) // Set player to survival before applying damage
 					((Player) entity).setGameMode(GameMode.SURVIVAL);
```
```suggestion
					player.setGameMode(GameMode.SURVIVAL);
```

EquipableMC replied:
Cannot be changed, it does not accept `player`, as seen below
![image](https://github.com/user-attachments/assets/a3cc9f4f-d4ea-4af1-9007-76cb0a554624)

Efnilite commented on src/main/java/ch/njol/skript/effects/EffKill.java:
```diff
@@ -26,40 +26,46 @@
+				return;
+			}
+
 			if (entity instanceof Damageable) {
```
```suggestion
			if (entity instanceof Damageable damageable) {
```

Efnilite commented on src/main/java/ch/njol/skript/effects/EffKill.java:
```diff
@@ -26,40 +26,46 @@
 				if (creative) // Set player to survival before applying damage
 					((Player) entity).setGameMode(GameMode.SURVIVAL);
-
 				HealthUtils.damage((Damageable) entity, HealthUtils.getMaxHealth((Damageable) entity) * 100); // just to make sure that it really dies >:)
```
```suggestion
				HealthUtils.damage(damageable, HealthUtils.getMaxHealth(damageable * 100); // just to make sure that it really dies >:)
```

Efnilite commented on src/main/java/ch/njol/skript/effects/EffKill.java:
```diff
@@ -26,40 +26,46 @@
-
 				HealthUtils.damage((Damageable) entity, HealthUtils.getMaxHealth((Damageable) entity) * 100); // just to make sure that it really dies >:)
 
 				if (creative) // Set creative player back to creative
```
line below this one can also have pattern matched var

EquipableMC commented:
I removed `ignoring the totem of undying` and removed setting the player to survival from creative, then back as that was unneeded, which is now making this a breaking change.

Efnilite reviewed (approved):
this will be a breaking change as the kill effect will now always kill, no matter if the player has a totem. in my opinion, this is expected behaviour, and therefore not a big deal

Moderocky commented on src/main/java/ch/njol/skript/effects/EffKill.java:
```diff
@@ -26,44 +26,35 @@
 		"Note: This effect does not set the entity's health to 0 (which causes issues), but damages the entity by 100 times its maximum health."})
 @Examples({"kill the player",
 		"kill all creepers in the player's world",
-		"kill all endermen, witches and bats"})
+		"kill all endermen, witches and bats",
+})
```
Why did you add a `,` and spaces at the end? Please keep it consistent for both brackets.

EquipableMC replied:
I thought I removed it, but I didn't realize I didn't, oopsies

Sparky200 commented on src/main/java/ch/njol/skript/effects/EffKill.java:
```diff
@@ -26,44 +26,35 @@
 	}
 	
-	// Absolutely make sure it dies
-	public static final int DAMAGE_AMOUNT = Integer.MAX_VALUE;
```
Just a general rule in API, you should add this back with a deprecated and scheduled for removal annotation. It's highly unlikely this is used by any consumers of the API, but nevertheless good practice in case it does, so the consumer has more opportunity to change their use case

EquipableMC replied:
That was legacy code, and I thought I should remove it, but it really didn't matter, since this is like the least important place to apply it.

Moderocky commented on src/main/java/ch/njol/skript/effects/EffKill.java:
```diff
@@ -44,24 +44,26 @@
-	// Absolutely make sure it dies
-	public static final int DAMAGE_AMOUNT = Integer.MAX_VALUE;
-	
 	@SuppressWarnings("null")
```
You need to actually do these changes rather than just clicking Resolve Conversation.

Moderocky commented on src/main/java/ch/njol/skript/effects/EffKill.java:
```diff
@@ -44,24 +44,26 @@
 	private Expression<Entity> entities;
+	private boolean IGNORE_TOTEM;
 	
 	@SuppressWarnings({"unchecked", "null"})
```
See above.
